File: backend/src/core/VaultManager.cpp
#include "VaultManager.hpp"
#include "FileSystemTool.hpp" // For directory creation
#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <random>
#include <cstring>
#include <openssl/sha.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// --- CONFIGURATION ---
// IMPORTANT: Update this path to where your compiled Kotlin JAR resides
const std::string JAR_PATH = "-Djava.class.path=cryptography-1.0.0-SNAPSHOT-uber.jar";
const std::string CRYPTO_DIR = "crypto_files";
const std::string PEPPER_FILE = CRYPTO_DIR + "/pepper.bin";
const std::string KEYSTORE_FILE = CRYPTO_DIR + "/key_store.vault";
const std::string VAULT_FILE = CRYPTO_DIR + "/secure_vault.dat";
const std::string KEY_ALIAS = "master_key";

// Static JVM members
JavaVM* VaultManager::jvm = nullptr;
JNIEnv* VaultManager::env = nullptr;

VaultManager::VaultManager() {
    m_pepper = loadOrGeneratePepper();
    initializeJVM();

    // 1. Initialize Classes and Objects
    if (!env) return;

    // Find Classes
    jclass clsKM = env->FindClass("com/personal/image_toolkit/KeyStoreManager");
    jclass clsKI = env->FindClass("com/personal/image_toolkit/KeyInitializer");
    jclass clsVault = env->FindClass("com/personal/image_toolkit/SecureJsonVault");

    if (!clsKM || !clsKI || !clsVault) {
        std::cerr << "Failed to find Kotlin classes. Check JAR path and classpath." << std::endl;
        checkJniException();
        return;
    }

    // Create Global References for Classes
    m_clsKeyStoreManager = (jclass)env->NewGlobalRef(clsKM);
    m_clsKeyInitializer = (jclass)env->NewGlobalRef(clsKI);
    m_clsSecureJsonVault = (jclass)env->NewGlobalRef(clsVault);

    // Instantiate KeyStoreManager
    jmethodID ctorKM = env->GetMethodID(m_clsKeyStoreManager, "<init>", "()V");
    jobject localKM = env->NewObject(m_clsKeyStoreManager, ctorKM);
    m_keystoreManagerObj = env->NewGlobalRef(localKM);

    // Instantiate KeyInitializer
    jmethodID ctorKI = env->GetMethodID(m_clsKeyInitializer, "<init>", "()V");
    jobject localKI = env->NewObject(m_clsKeyInitializer, ctorKI);
    m_keyInitializerObj = env->NewGlobalRef(localKI);
    
    checkJniException();
}

VaultManager::~VaultManager() {
    if (env) {
        if (m_keystoreManagerObj) env->DeleteGlobalRef(m_keystoreManagerObj);
        if (m_keyInitializerObj) env->DeleteGlobalRef(m_keyInitializerObj);
        if (m_vaultObj) env->DeleteGlobalRef(m_vaultObj);
        if (m_keystoreObj) env->DeleteGlobalRef(m_keystoreObj);
        if (m_secretKeyObj) env->DeleteGlobalRef(m_secretKeyObj);
        
        if (m_clsKeyStoreManager) env->DeleteGlobalRef(m_clsKeyStoreManager);
        if (m_clsKeyInitializer) env->DeleteGlobalRef(m_clsKeyInitializer);
        if (m_clsSecureJsonVault) env->DeleteGlobalRef(m_clsSecureJsonVault);
    }
}

void VaultManager::initializeJVM() {
    if (jvm != nullptr) {
        // Attach current thread if JVM exists
        jvm->AttachCurrentThread((void**)&env, nullptr);
        return;
    }

    JavaVMInitArgs vm_args;
    JavaVMOption options[1];
    
    // Set Classpath
    // In C++, we need to construct a char* for the option
    char* classpath_opt = new char[JAR_PATH.length() + 1];
    std::strcpy(classpath_opt, JAR_PATH.c_str());
    options[0].optionString = classpath_opt;

    vm_args.version = JNI_VERSION_1_8;
    vm_args.nOptions = 1;
    vm_args.options = options;
    vm_args.ignoreUnrecognized = JNI_FALSE;

    jint rc = JNI_CreateJavaVM(&jvm, (void**)&env, &vm_args);
    delete[] classpath_opt;

    if (rc != JNI_OK) {
        std::cerr << "Failed to create JVM: " << rc << std::endl;
        exit(1);
    }
}

void VaultManager::shutdownJVM() {
    if (jvm) {
        jvm->DestroyJavaVM();
        jvm = nullptr;
        env = nullptr;
    }
}

// --- Helpers ---

jcharArray VaultManager::stringToCharArray(const std::string& str) {
    // Determine length (UTF-16 chars)
    // Simple ASCII conversion for passwords/paths
    jcharArray charArr = env->NewCharArray(str.length());
    std::vector<jchar> buffer(str.length());
    for(size_t i=0; i<str.length(); ++i) buffer[i] = (jchar)str[i];
    
    env->SetCharArrayRegion(charArr, 0, str.length(), buffer.data());
    return charArr;
}

std::string VaultManager::jstringToString(jstring jStr) {
    if (!jStr) return "";
    const char* chars = env->GetStringUTFChars(jStr, nullptr);
    std::string ret(chars);
    env->ReleaseStringUTFChars(jStr, chars);
    return ret;
}

void VaultManager::checkJniException() {
    if (env->ExceptionCheck()) {
        env->ExceptionDescribe(); // Prints to stderr
        env->ExceptionClear();
        throw std::runtime_error("Java Exception Occurred");
    }
}

// --- Business Logic Wrappers ---

void VaultManager::loadKeystore(const std::string& keystore_path, const std::string& keystore_pass) {
    jmethodID mid = env->GetMethodID(m_clsKeyStoreManager, "loadKeyStore", "(Ljava/lang/String;[C)Ljava/security/KeyStore;");
    if (!mid) return;

    jstring jPath = env->NewStringUTF(keystore_path.c_str());
    jcharArray jPass = stringToCharArray(keystore_pass);

    jobject localKS = env->CallObjectMethod(m_keystoreManagerObj, mid, jPath, jPass);
    checkJniException();

    if (m_keystoreObj) env->DeleteGlobalRef(m_keystoreObj);
    m_keystoreObj = env->NewGlobalRef(localKS);
    
    std::cout << "Keystore loaded successfully via JNI." << std::endl;
}

bool VaultManager::containsAlias(const std::string& key_alias) {
    if (!m_keystoreObj) throw std::runtime_error("Keystore not loaded");

    // We can call KeyStore.containsAlias directly or use KeyStoreManager wrapper if it exposes it
    // Assuming wrapper relies on underlying KeyStore object, let's call KeyStore method directly for simplicity
    jclass clsKS = env->GetObjectClass(m_keystoreObj);
    jmethodID mid = env->GetMethodID(clsKS, "containsAlias", "(Ljava/lang/String;)Z");
    
    jstring jAlias = env->NewStringUTF(key_alias.c_str());
    jboolean result = env->CallBooleanMethod(m_keystoreObj, mid, jAlias);
    checkJniException();
    
    return (bool)result;
}

void VaultManager::createKeyIfMissing(const std::string& key_alias, const std::string& keystore_path, const std::string& keystore_pass) {
    jmethodID mid = env->GetMethodID(m_clsKeyInitializer, "initializeKeystore", "(Ljava/lang/String;Ljava/lang/String;[C[C)V");
    if (!mid) return;

    jstring jPath = env->NewStringUTF(keystore_path.c_str());
    jstring jAlias = env->NewStringUTF(key_alias.c_str());
    jcharArray jPass = stringToCharArray(keystore_pass);

    env->CallVoidMethod(m_keyInitializerObj, mid, jPath, jAlias, jPass, jPass);
    checkJniException();

    // Reload logic
    loadKeystore(keystore_path, keystore_pass);
}

void VaultManager::getSecretKey(const std::string& key_alias, const std::string& key_pass) {
    if (!m_keystoreObj) throw std::runtime_error("Keystore not loaded");

    jmethodID mid = env->GetMethodID(m_clsKeyStoreManager, "getSecretKey", "(Ljava/security/KeyStore;Ljava/lang/String;[C)Ljavax/crypto/SecretKey;");
    
    jstring jAlias = env->NewStringUTF(key_alias.c_str());
    jcharArray jPass = stringToCharArray(key_pass);

    jobject localKey = env->CallObjectMethod(m_keystoreManagerObj, mid, m_keystoreObj, jAlias, jPass);
    checkJniException();

    if (m_secretKeyObj) env->DeleteGlobalRef(m_secretKeyObj);
    m_secretKeyObj = env->NewGlobalRef(localKey);
    
    std::cout << "SecretKey retrieved via JNI." << std::endl;
}

void VaultManager::initVault(const std::string& vault_file_path) {
    if (!m_secretKeyObj) throw std::runtime_error("SecretKey not loaded");

    jmethodID ctor = env->GetMethodID(m_clsSecureJsonVault, "<init>", "(Ljavax/crypto/SecretKey;Ljava/lang/String;)V");
    jstring jPath = env->NewStringUTF(vault_file_path.c_str());

    jobject localVault = env->NewObject(m_clsSecureJsonVault, ctor, m_secretKeyObj, jPath);
    checkJniException();

    if (m_vaultObj) env->DeleteGlobalRef(m_vaultObj);
    m_vaultObj = env->NewGlobalRef(localVault);
    
    std::cout << "SecureJsonVault initialized via JNI." << std::endl;
}

void VaultManager::saveData(const std::string& json_string) {
    if (!m_vaultObj) throw std::runtime_error("Vault not initialized");

    jmethodID mid = env->GetMethodID(m_clsSecureJsonVault, "saveData", "(Ljava/lang/String;)V");
    jstring jData = env->NewStringUTF(json_string.c_str());

    env->CallVoidMethod(m_vaultObj, mid, jData);
    checkJniException();
}

std::string VaultManager::loadData() {
    if (!m_vaultObj) throw std::runtime_error("Vault not initialized");

    jmethodID mid = env->GetMethodID(m_clsSecureJsonVault, "loadData", "()Ljava/lang/String;");
    jstring jResult = (jstring)env->CallObjectMethod(m_vaultObj, mid);
    
    // Check if exception occurred (e.g., file not found)
    if (env->ExceptionCheck()) {
        env->ExceptionClear(); 
        return "{}"; // Default to empty json
    }

    return jstringToString(jResult);
}

// --- C++ Side Logic (Pepper, Hashing, JSON) ---
// Note: This remains largely pure C++ because it interacts with Python-style logic 
// not the Kotlin vault itself, except for the saveData call.

std::string VaultManager::loadOrGeneratePepper() {
    FSETool::createDirectory(CRYPTO_DIR);
    std::ifstream file(PEPPER_FILE);
    if (file.is_open()) {
        std::string pepper;
        std::getline(file, pepper);
        return pepper;
    } else {
        std::random_device rd;
        std::vector<unsigned char> random_bytes(32);
        for(unsigned char& b : random_bytes) b = static_cast<unsigned char>(rd());
        std::stringstream ss;
        for(unsigned char b : random_bytes) ss << std::hex << std::setw(2) << std::setfill('0') << (int)b;
        std::string pepper = ss.str();
        std::ofstream outfile(PEPPER_FILE);
        outfile << pepper;
        return pepper;
    }
}

std::string VaultManager::hashPassword(const std::string& raw_pass, const std::string& salt) {
    std::string combined = raw_pass + salt + m_pepper;
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, combined.c_str(), combined.length());
    SHA256_Final(hash, &sha256);
    std::stringstream ss;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    return ss.str();
}

void VaultManager::saveAccountCredentials(const std::string& account_name, const std::string& raw_password) {
    std::random_device rd;
    std::vector<unsigned char> random_bytes(16);
    for(unsigned char& b : random_bytes) b = static_cast<unsigned char>(rd());
    std::stringstream salt_ss;
    for(unsigned char b : random_bytes) salt_ss << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    std::string salt = salt_ss.str();

    std::string hashed_password = hashPassword(raw_password, salt);

    json data;
    data["account_name"] = account_name;
    data["hashed_password"] = hashed_password;
    data["salt"] = salt;

    saveData(data.dump());
}

std::map<std::string, std::string> VaultManager::loadAccountCredentials() {
    std::string json_str = loadData();
    if (json_str == "{}" || json_str.empty()) return {};
    
    try {
        json j = json::parse(json_str);
        std::map<std::string, std::string> res;
        res["account_name"] = j["account_name"];
        res["hashed_password"] = j["hashed_password"];
        res["salt"] = j["salt"];
        return res;
    } catch (...) {
        return {};
    }
}

void VaultManager::updateAccountPassword(const std::string& account_name, const std::string& new_raw_pass) {
    // 1. Load Data
    std::string old_json = loadData();
    json j_old = json::parse(old_json);

    // 2. Clear References (Important so file locks are released by JVM GC)
    env->DeleteGlobalRef(m_vaultObj); m_vaultObj = nullptr;
    env->DeleteGlobalRef(m_keystoreObj); m_keystoreObj = nullptr;
    env->DeleteGlobalRef(m_secretKeyObj); m_secretKeyObj = nullptr;
    
    // Force GC? (Optional, JNI doesn't guarantee immediate GC)
    
    // 3. Delete Files
    FileDeleter::deletePath(KEYSTORE_FILE);
    FileDeleter::deletePath(VAULT_FILE);

    // 4. Create New
    createKeyIfMissing(KEY_ALIAS, KEYSTORE_FILE, new_raw_pass);
    getSecretKey(KEY_ALIAS, new_raw_pass);
    initVault(VAULT_FILE);

    // 5. Update Data
    std::random_device rd;
    std::vector<unsigned char> random_bytes(16);
    for(unsigned char& b : random_bytes) b = static_cast<unsigned char>(rd());
    std::stringstream new_salt_ss;
    for(unsigned char b : random_bytes) new_salt_ss << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    std::string new_salt = new_salt_ss.str();

    j_old["hashed_password"] = hashPassword(new_raw_pass, new_salt);
    j_old["salt"] = new_salt;

    saveData(j_old.dump());
}

File: backend/src/core/vault_manager.py
import os
import json
import jpype
import hashlib

try:
    import backend.src.utils.definitions as udef
except:
    import src.utils.definitions as udef

from jpype.types import JArray, JChar


class VaultManager:
    """
    A Python wrapper to manage the SecureJsonVault by calling
    the compiled Java/Kotlin code via Jpype.
    """

    @staticmethod
    def _load_or_generate_pepper():
        """
        Checks for the pepper file. If it doesn't exist, it generates a new
        secure pepper and saves it.
        """
        # Ensure the directory for crypto files exists
        os.makedirs(udef.CRYPTO_DIR, exist_ok=True)

        if os.path.exists(udef.PEPPER_FILE):
            print(f"Loading existing pepper from: {udef.PEPPER_FILE}")
            with open(udef.PEPPER_FILE, "r") as f:
                pepper = f.read().strip()
                if not pepper:
                    raise ValueError(
                        f"Pepper file is empty: {udef.PEPPER_FILE}. Delete it to regenerate."
                    )
                return pepper
        else:
            print(
                f"Pepper file not found. Generating new pepper at: {udef.PEPPER_FILE}"
            )
            # Generate a strong, random pepper (32 bytes = 64 hex characters)
            pepper = os.urandom(32).hex()

            # Save the new pepper to the file
            with open(udef.PEPPER_FILE, "w") as f:
                f.write(pepper)

            # Set restrictive permissions (read-only for owner, if supported)
            try:
                os.chmod(udef.PEPPER_FILE, 0o400)
            except OSError:
                # Handle systems that don't support chmod (like some Windows systems)
                print(
                    "Warning: Could not set restrictive file permissions on pepper file."
                )

            return pepper

    def __init__(self, bc_provider_path: str = None):
        """
        Initializes the wrapper, starts the JVM, and loads the pepper.
        """
        # --- Load or generate the secret pepper first ---
        self.PEPPER = self._load_or_generate_pepper()

        # Check if JVM is already running to avoid crash on re-initialization
        if not jpype.isJVMStarted():
            # NOTE: Ensure udef.JAR_FILE points to the new Gradle build output
            # (e.g., backend/cryptography/build/libs/cryptography-1.0.0-SNAPSHOT-uber.jar)
            classpath = [udef.JAR_FILE]
            if bc_provider_path:
                classpath.append(f"{bc_provider_path}/*")

            print(f"Starting JVM with classpath: {classpath}")
            jpype.startJVM(classpath=classpath)

        try:
            # Load Kotlin classes (Same package structure as Java)
            KeyStoreManagerClass = jpype.JClass(
                "com.personal.image_toolkit.KeyStoreManager"
            )
            KeyInitializerClass = jpype.JClass(
                "com.personal.image_toolkit.KeyInitializer"
            )
            self.SecureJsonVault = jpype.JClass(
                "com.personal.image_toolkit.SecureJsonVault"
            )

            self.keystore_manager = KeyStoreManagerClass()
            self.key_initializer = KeyInitializerClass()

        except jpype.JException as e:
            print("\n--- ERROR ---")
            print(f"Could not find Java/Kotlin class: {e}")
            print(
                "Ensure the JAR file path in definitions.py is correct and the project is built."
            )
            raise

        self.JString = jpype.JClass("java.lang.String")

        self.keystore = None
        self.secret_key = None
        self.vault = None

        # --- THIS DICTIONARY WILL HOLD DECRYPTED API CREDENTIALS ---
        self.api_credentials = {}

    def _to_char_array(self, py_string: str) -> "JArray[JChar]":
        """Helper to convert a Python string to a Java char[]."""
        return self.JString(py_string).toCharArray()

    def load_keystore(self, keystore_path: str, keystore_pass: str):
        """
        Loads the Java KeyStore from a file using the KeyStoreManager.
        """
        try:
            print(f"Loading keystore: {keystore_path}")
            self.keystore = self.keystore_manager.loadKeyStore(
                keystore_path, self._to_char_array(keystore_pass)
            )
            print("Keystore loaded successfully.")
        except Exception as e:
            print(f"Java Error loading keystore: {e}")
            raise

    def contains_alias(self, key_alias: str) -> bool:
        """
        Checks if the loaded KeyStore contains an entry for the given alias.
        """
        if self.keystore is None:
            raise ValueError("Keystore is not loaded. Call load_keystore() first.")

        try:
            return self.keystore.containsAlias(key_alias)
        except Exception as e:
            print(f"Java Error checking alias: {e}")
            raise

    def create_key_if_missing(
        self, key_alias: str, keystore_path: str, keystore_pass: str
    ):
        """
        Uses the KeyInitializer to ensure the keystore exists and contains
        the required secret key.
        """
        print(
            f"Checking/Initializing KeyStore at {keystore_path} for alias '{key_alias}'..."
        )

        try:
            # Delegate the check-and-create logic to the KeyInitializer.
            # This is safer as it handles the file IO and key generation atomically on the JVM side.
            self.key_initializer.initializeKeystore(
                keystore_path,
                key_alias,
                self._to_char_array(keystore_pass),
                self._to_char_array(
                    keystore_pass
                ),  # Using same pass for store and key for simplicity
            )

            # CRITICAL: After KeyInitializer potentially modifies the file on disk,
            # we must reload it into our Python memory object to get the new key.
            self.load_keystore(keystore_path, keystore_pass)

        except Exception as e:
            print(f"Java Error in KeyInitializer: {e}")
            raise

    def get_secret_key(self, key_alias: str, key_pass: str):
        """
        Retrieves the AES SecretKey from the loaded KeyStore.
        """
        if self.keystore is None:
            raise ValueError("Keystore is not loaded. Call load_keystore() first.")

        try:
            print(f"Retrieving secret key for alias: {key_alias}")
            self.secret_key = self.keystore_manager.getSecretKey(
                self.keystore, key_alias, self._to_char_array(key_pass)
            )
            if self.secret_key is None:
                # This should only happen if the key type is wrong or password is bad
                raise ValueError(
                    f"No secret key found for alias '{key_alias}' or wrong password."
                )
            print("SecretKey retrieved.")
        except Exception as e:
            print(f"Java Error getting secret key: {e}")
            raise

    def init_vault(self, vault_file_path: str):
        """
        Initializes the SecureJsonVault with the retrieved SecretKey.
        """
        if self.secret_key is None:
            raise ValueError("Secret key is not loaded. Call get_secret_key() first.")

        print(f"Initializing secure vault at: {vault_file_path}")
        self.vault = self.SecureJsonVault(self.secret_key, vault_file_path)
        print("Vault initialized.")

    def save_data(self, json_string: str):
        """
        Saves a JSON string to the encrypted vault.
        """
        if self.vault is None:
            raise ValueError("Vault is not initialized. Call init_vault() first.")

        try:
            self.vault.saveData(json_string)
            print("Data saved successfully.")
        except Exception as e:
            print(f"Java Error saving data: {e}")
            raise

    def load_data(self) -> str:
        """
        Loads and decrypts the JSON string from the vault.
        """
        if self.vault is None:
            raise ValueError("Vault is not initialized. Call init_vault() first.")

        try:
            print("Loading data from vault...")
            decrypted_data = self.vault.loadData()
            print("Data loaded and decrypted successfully.")
            decrypted_json_string = str(decrypted_data)
            return decrypted_json_string

        except Exception as e:
            # Handle empty vault file as an empty JSON object string
            if (
                "file not found" in str(e).lower()
                or "vault file not found" in str(e).lower()
            ):
                print("Vault file is empty or not found. Returning empty JSON object.")
                return "{}"

            print(f"Java Error loading data: {e}")
            print("This may be due to a wrong key or file tampering.")
            raise

    def update_account_password(self, account_name: str, new_raw_pass: str):
        """
        Updates the master password by replacing the keystore and re-encrypting
        the vault data with a new key derived from the new password.

        This method is non-destructive (preserves data).
        """
        print("Starting master password update process (data preserving)...")

        # 1. Retrieve ALL data from the currently open vault (encrypted with OLD key)
        try:
            # This loads everything, including the account hash/salt
            old_vault_content_json = self.load_data()
            old_vault_content = json.loads(old_vault_content_json)
        except Exception as e:
            raise RuntimeError(f"Failed to load data from old vault before reset: {e}")

        # 2. Shutdown JVM to release file locks before deleting files
        # NOTE: JPype cannot restart the JVM in the same process.
        # This method assumes the script will exit or the architecture supports this constraint.
        # For a long-running app, replacing files without full JVM restart is safer.
        # However, Java file locks might persist if streams aren't closed (Kotlin 'use' block handles this).

        # 3. Delete Old Keystore and Vault
        if os.path.exists(udef.KEYSTORE_FILE):
            os.remove(udef.KEYSTORE_FILE)
            print(f"Deleted old KeyStore: {udef.KEYSTORE_FILE}")

        if os.path.exists(udef.VAULT_FILE):
            os.remove(udef.VAULT_FILE)
            print(f"Deleted old Vault file: {udef.VAULT_FILE}")

        # 4. Re-initialize internal objects (JVM remains running)
        # We don't need to call __init__ again because JVM is already up.
        # We just need to clear the python state references.
        self.keystore = None
        self.secret_key = None
        self.vault = None

        # 5. Create new KeyStore/Key Entry/Vault with the NEW password
        # Uses the updated create_key_if_missing logic which uses KeyInitializer
        self.create_key_if_missing(udef.KEY_ALIAS, udef.KEYSTORE_FILE, new_raw_pass)

        # Get the new key reference
        self.get_secret_key(udef.KEY_ALIAS, new_raw_pass)
        self.init_vault(udef.VAULT_FILE)

        # 6. Re-hash and update the account credentials (hash/salt) within the data
        new_salt = os.urandom(16).hex()
        password_combined = (new_raw_pass + new_salt + self.PEPPER).encode("utf-8")
        new_hashed_password = hashlib.sha256(password_combined).hexdigest()

        old_vault_content.update(
            {
                "account_name": account_name,
                "hashed_password": new_hashed_password,
                "salt": new_salt,
            }
        )

        # 7. Encrypt and save the updated data (with new hash/salt) to the new vault
        new_json_string = json.dumps(old_vault_content)
        self.save_data(new_json_string)

        print("Master password update complete. Data preserved and re-encrypted.")

    def shutdown(self):
        """Shuts down the JVM if it's running."""
        if jpype.isJVMStarted():
            print("Shutting down JVM...")
            jpype.shutdownJVM()
            print("JVM shut down.")

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.shutdown()
        return False

    def save_account_credentials(self, account_name: str, raw_password: str):
        """
        Hashes and salts the raw password with the loaded pepper, then saves the
        account name, resulting hash, and salt to the encrypted vault.
        """
        # 1. Generate a secure, unique salt (16 bytes)
        salt = os.urandom(16).hex()

        # 2. Combine all security components for hashing
        password_combined = (raw_password + salt + self.PEPPER).encode("utf-8")

        # 3. Hash the combined string (SHA-256 is used for simplicity)
        hashed_password = hashlib.sha256(password_combined).hexdigest()

        data_to_save = {
            "account_name": account_name,
            "hashed_password": hashed_password,
            "salt": salt,
        }

        json_string = json.dumps(data_to_save)

        print(f"Saving credentials for account: {account_name}")
        self.save_data(json_string)  # Uses the existing saveData() method

    def load_account_credentials(self) -> dict:
        """
        Loads, decrypts, and parses the account name, hashed password, and salt
        from the vault.

        :return: A dictionary containing the loaded credentials.
        """
        decrypted_json_string = self.load_data()  # Uses the existing loadData() method

        try:
            loaded_data = json.loads(decrypted_json_string)

            # Ensure the required keys are present
            required_keys = ["account_name", "hashed_password", "salt"]
            if not all(key in loaded_data for key in required_keys):
                # Handle empty init case
                if loaded_data == {}:
                    return {}
                raise KeyError(
                    f"Decrypted JSON is missing one of the required keys: {required_keys}"
                )

            return loaded_data

        except (json.JSONDecodeError, KeyError) as e:
            print(f"Error parsing loaded data: {e}")
            raise ValueError("The vault file contains invalid or corrupted JSON data.")
